Add ldap server example (#105)
Mirantis/mke-docs: content/docs/tutorials/authentication-provider-setup/setting-up-openldap-as-an-ldap-provider.md

---
title: Setting up OpenLDAP as an LDAP provider
weight: 1
---

To create a basic OpenLDAP server to test [LDAP](../../../../docs/operations/authentication/ldap) with MKE4:

> > **Note** To the the OpenLDAP server you must have [Docker and docker compose](https://docs.docker.com/engine/install/) installed on your system.

Reviewer: ```suggestion
>{{< callout type="info" >}}
  To run the OpenLDAP server you must have [Docker and docker compose](https://docs.docker.com/engine/install/) installed on your system.
{{< /callout >}}
```

    # o: Example Company
    # dc: example

    dn: ou=People,dc=example,dc=org
    objectClass: organizationalUnit
    ou: People

    dn: cn=jane,ou=People,dc=example,dc=org
    objectClass: person
    objectClass: inetOrgPerson
    sn: doe
    cn: jane
    mail: janedoe@example.com
    userpassword: foo

    dn: cn=john,ou=People,dc=example,dc=org
    objectClass: person
    objectClass: inetOrgPerson
    sn: doe
    cn: john
    mail: johndoe@example.com
    userpassword: bar

    # Group definitions.

    dn: ou=Groups,dc=example,dc=org
    objectClass: organizationalUnit
    ou: Groups

    dn: cn=admins,ou=Groups,dc=example,dc=org
    objectClass: groupOfNames  dns:
        lameduck:
          enabled: true
          duration: "7s"
    cn: admins
    member: cn=john,ou=People,dc=example,dc=org
    member: cn=jane,ou=People,dc=example,dc=org

    dn: cn=developers,ou=Groups,dc=example,dc=org
    objectClass: groupOfNames
    cn: developers
    member: cn=jane,ou=People,dc=example,dc=org
   ```

   The `config-ldap.ldif` file will be loaded into the Docker container for the LDAP server, from where the server will run the file to create a number of default users.

> 2. Create the docker compose file called `docker-compose.yml`, containing the following content:

Reviewer: ```suggestion
2. Create a Docker Compose file named `docker-compose.yml` that contains the following content:
```

       # Copying is required because the entrypoint modifies the *.ldif files.
       # For verbose output, use:  command: ["--copy-service", "--loglevel", "debug"]
       command: ["--copy-service", "--loglevel", "debug"]
       environment:
         # Required if using LDAPS:
         # Since Dex doesn't use a client TLS certificate, downgrade from "demand" to "try".
         LDAP_TLS_VERIFY_CLIENT: try
       # The hostname is required if using LDAPS with certificate validation.
       # In Dex, use the same hostname (with port) for `connectors[].config.host`.
       #hostname: YOUR-HOSTNAME
       #
       # https://github.com/osixia/docker-openldap#seed-ldap-database-with-ldif
       # Option 1: Add custom seed file -> mount to         /container/service/slapd/assets/config/bootstrap/ldif/custom/
       # Option 2: Overwrite default seed file -> mount to  /container/service/slapd/assets/config/bootstrap/ldif/
       volumes:
         - ./config-ldap.ldif:/container/service/slapd/assets/config/bootstrap/ldif/custom/config-ldap.ldif
       ports:
         - 389:389
         - 636:636
   ```

3. Start the OpenLDAP server:

   ```bash
   docker compose up
   ```

>    The Docker Compose file copies in the `config-ldap.ldif` configuration file and uses it in setting up the LDAP server which exposes the LDAP service on port `389` for non-TLS and `636` for TLS.

Reviewer: ```suggestion
   The Docker Compose file copies in the `config-ldap.ldif` configuration file and uses it to set up the LDAP server, which exposes the LDAP service on port `389` for non-TLS and `636` for TLS.
```

## Configure MKE to use LDAP

An example MKE4 configuration for LDAP is shown below, for the exemplified LDAP server:

```yaml
authentication:
  enabled: true
  ldap:
    enabled: true
    host: localhost:389
    insecureNoSSL: true
    bindDN: cn=admin,dc=mirantis,dc=org
    bindPW: admin
    usernamePrompt: Email Address
    userSearch:
      baseDN: ou=People,dc=mirantis,dc=org
      filter: "(objectClass=person)"
      username: mail
      idAttr: DN
      emailAttr: mail
      nameAttr: cn
```

Run the `mkectl apply` command using the configuration file to apply the configuration to MKE.

```bash
mkectl apply -f config.yaml
```

## Test authentication flow

1. Navigate to the MKE dashboard: `https://{MKE hostname}`

> 2. Select **Log in with LDAP**. This will redirect you to the LDAP login page.

Reviewer: ```suggestion
2. Select **Log in with LDAP** to access the LDAP login page.
```

---
title: Setting up OpenLDAP as an LDAP provider
weight: 1
---

To create an OpenLDAP server to test [LDAP](../../../../docs/operations/authentication/ldap) with MKE4:

> ## Set up an LDAP server

Reviewer: The title of the topic is sufficient as a lead-in. Let's drop this heading.

> > **Note** Running the development server requires [Docker and docker compose](https://docs.docker.com/engine/install/) to be installed on your machine. Docker compose is now part of docker

Reviewer: Edit note to:

{{< callout type="info" >}}
To run the development server you must have [Docker and docker compose](https://docs.docker.com/engine/install/) installed on your system.
{{< /callout >}}

Author: Changed but I used `OpenLDAP server` instead of development server.

> 1. Create the initial config file:
> 
>    Create a file called `config-ldap.ldif` with the following content:

Reviewer: ```suggestion
1. Create a file called `config-ldap.ldif` with the following content:
```

   ```ldif
    # dn: dc=example,dc=org
    # objectClass: dcObject
    # objectClass: organization
    # o: Example Company
    # dc: example

    dn: ou=People,dc=example,dc=org
    objectClass: organizationalUnit
    ou: People

    dn: cn=jane,ou=People,dc=example,dc=org
    objectClass: person
    objectClass: inetOrgPerson
    sn: doe
    cn: jane
    mail: janedoe@example.com
    userpassword: foo

    dn: cn=john,ou=People,dc=example,dc=org
    objectClass: person
    objectClass: inetOrgPerson
    sn: doe
    cn: john
    mail: johndoe@example.com
    userpassword: bar

    # Group definitions.

    dn: ou=Groups,dc=example,dc=org
    objectClass: organizationalUnit
    ou: Groups

    dn: cn=admins,ou=Groups,dc=example,dc=org
    objectClass: groupOfNames  dns:
        lameduck:
          enabled: true
          duration: "7s"
    cn: admins
    member: cn=john,ou=People,dc=example,dc=org
    member: cn=jane,ou=People,dc=example,dc=org

    dn: cn=developers,ou=Groups,dc=example,dc=org
    objectClass: groupOfNames
    cn: developers
    member: cn=jane,ou=People,dc=example,dc=org
   ```

>    The config-ldap.ldif file will be loaded into the Docker container for the LDAP server, from where the server will run the file to create a number of default users.

Reviewer: ```suggestion
   The `config-ldap.ldif file` will be loaded into the Docker container for the LDAP server, from where the server will run the file to create a number of default users.
```

> 2. Create the docker compose file:
>    Create a file called `docker-compose.yml` with the following content:

Reviewer: ```suggestion
2. Create a docker compose file called `docker-compose.yml`, containing the following content:
```

   services:
     ldap:
       image: osixia/openldap:1.4.0
       # Copying is required because the entrypoint modifies the *.ldif files.
       # For verbose output, use:  command: ["--copy-service", "--loglevel", "debug"]
       command: ["--copy-service", "--loglevel", "debug"]
       environment:
         # Required if using LDAPS:
         # Since Dex doesn't use a client TLS certificate, downgrade from "demand" to "try".
         LDAP_TLS_VERIFY_CLIENT: try
       # The hostname is required if using LDAPS with certificate validation.
       # In Dex, use the same hostname (with port) for `connectors[].config.host`.
       #hostname: YOUR-HOSTNAME
       #
       # https://github.com/osixia/docker-openldap#seed-ldap-database-with-ldif
       # Option 1: Add custom seed file -> mount to         /container/service/slapd/assets/config/bootstrap/ldif/custom/
       # Option 2: Overwrite default seed file -> mount to  /container/service/slapd/assets/config/bootstrap/ldif/
       volumes:
         - ./config-ldap.ldif:/container/service/slapd/assets/config/bootstrap/ldif/custom/config-ldap.ldif
       ports:
         - 389:389
         - 636:636
   ```

> 3. Startup the server:
> 
>    The server can now be started with the command

Reviewer: ```suggestion
3. Start the server:
```

   ```bash
   docker compose up
   ```

>    This docker compose file will copy in the `config-ldap.ldif` configuration file and use it when setting up the LDAP server. The server will expose the LDAP service on port 389 for non-tls and 636 for tls.

Reviewer: ```suggestion
   The Docker Compose file copies in the `config-ldap.ldif` configuration file and uses it in setting up the LDAP server, which exposes the LDAP service on port `389` for non-TLS and on `636` for TLS.
```

## Configure MKE to use LDAP

> An example MKE4 configuration for LDAP is shown below. This configuration is for the example LDAP server.

Reviewer: ```suggestion
An example configuration for LDAP is shown below, for the exemplified LDAP server:
```

```yaml
authentication:
  enabled: true
  ldap:
    enabled: true
    host: localhost:389
    insecureNoSSL: true
    bindDN: cn=admin,dc=mirantis,dc=org
    bindPW: admin
    usernamePrompt: Email Address
    userSearch:
      baseDN: ou=People,dc=mirantis,dc=org
      filter: "(objectClass=person)"
      username: mail
      idAttr: DN
      emailAttr: mail
      nameAttr: cn
```

> Run the `mkectl apply` command with the configuration file to apply the configuration to MKE.

Reviewer: ```suggestion
Run the `mkectl apply` command using the configuration file to apply the configuration to MKE.
```
